Add workbench and nametag to rename items

octacian opened the pull request:
![Screenshot](http://i.imgur.com/Kf07POyg.png)

![Formspec Screenshot](https://cloud.githubusercontent.com/assets/15371866/23389894/acf991fe-fd1f-11e6-8466-a02ac9a7b84a.png)
Note: Formspec is the same size as the bookshelf formspec.

This PR introduces a new Workbench node and nametag item. These two items combined allow any item, node, or tool's description to be changed using the new key-based ItemStack meta.

The Workbench has a fairly simple formspec, with three slots. The first slot in the input item that you want to rename (can be anything but a nametag, max 1 item), the second slot contains only nametags, and the third and final slow is the output item given after clicking "Rename." If any of the items are missing or there is no room in the output slot when rename is clicked, an error message is displayed above the text field. When the proper item is inserted, the error message is automatically removed.

**Recipes:**
```lua
-- Workbench (gives 1 workbench)
{'default:steel_ingot', 'default:steel_ingot', 'default:steel_ingot'}
{'group:wood', '', 'group:wood'}
{'group:wood', '', 'group:wood'}

-- Nametag (gives 4 nametags)
{ "default:copper_ingot" }
```

The workbench item itself is rather complex with nearly 200 lines of code. With that in mind, there are likely many areas where things could be improved. Considering the complexity, it might even be better to move at least the workbench to a new file (not unlike the furnace), which I am open to do if requested.

sofar commented:
Please add a screenshot of the formspec display.

C1ffisme commented:
@octacian I would make nametags use string and paper, though, since name-tags don't make a lot of sense to me. Otherwise, this looks good.

octacian commented:
I had considered using string and paper, but it is nice to give copper a use, plus, the workbench texture it would make it appear more as though the nametag would be metal. That's why I used copper and a more coppery texture.

I may change it yet, but there is also a chance that I will attempt to do this in the craftgrid as sofar prefers it and it would actually make more sense seeing as tools are repaired in the craftgrid. So, we'll see.

paramat commented:
Does this workbench do anything other than rename items?
What are the uses for renaming in MTGame? Is it for renaming keys so they become individual?
I would prefer a node since this complex nodebox is not essential (but i admit looks nice).

sofar commented:
BTW, it's "McWorkyWorkbenchFace"

octacian commented:
No, the workbench does not do anything else. I was thinking that if this PR were merged, tool repair could be moved from the crafting grid to the workbench. Though it might be a bit complicated, overall it would make things more interesting. The main reason is for renaming keys, yes. Though it could prove pretty interesting for map creators. Sofar mentioned that he'd prefer a node as well, I'll see if I can find some textures. Personally I prefer the nodebox, I don't think MTG has enough variety in areas like that (making things less interesting).

Sofar and I have also been discussing implementing this in the crafting grid using a `renamable` group, however, I see some issues with that in that I prefer not to limit the things which can be renamed, but rather allow using a group to prevent renaming in something like the workbench. There are no good alternatives with the crafting grid method aside from groups, so it's a hard call as iterating through items to add the group if it's not already set is terrible. So, with that in mind, I'll open a second PR using a limited groups method tomorrow, then we can discuss whether or not to make a point of trying to remove the limitation one way or another.

paramat commented:
Thanks, the only other issue is the font used, seems small and does not match the font used in inventory.

octacian commented:
@paramat I didn't hardcode the font or the size (is that even possible?). It is just the font that I set for my local MT installation. To anybody else, the formspec would use the default font or whatever font they had specified in their `minetest.conf`.

BTW, are there any people who would specifically prefer the crafting grid method over this? IK sofar does, but I don't know about anybody else.

paramat commented:
No problem i don't understand how fonts work.
I'm considering the crafting grid approach.

rubenwardy commented:
This should be a new mod, and it should be something like "renamer workbench" or "nametag workbench"

Fixer-007 commented:
Works nice so far, I noticed one thing: if you rename it to looots of text, then tooltip will obscure renamed node itself, if you rename even more - can obscure entire screen, I managed to get out the nodes though :)
Examples:
https://i.imgur.com/6YHifNE.png - no problem (few text)
https://i.imgur.com/YelhuWW.png - problem (more text)
https://i.imgur.com/xCIju4Q.png - problem (README.txt...)

C1ffisme commented:
@octacian 
> I had considered using string and paper, but it is nice to give copper a use, plus, the workbench texture it would make it appear more as though the nametag would be metal. That's why I used copper and a more coppery texture.

Okay, I can agree with that.

Fixer-007 commented:
Workbench fits nice as furniture https://i.imgur.com/bUazaRr.png

SmallJoker commented:
> Considering the complexity, it might even be better to move at least the workbench to a new file (not unlike the furnace), which I am open to do if requested.

Yes, the `nodes.lua` file is already quite long. I would like to see it in a different file.

@rubenwardy This node does not add any API or changes the map generation, so I think it's fine to keep it in default.

SmallJoker commented on mods/default/nodes.lua:
```diff
@@ -1761,6 +1762,218 @@ minetest.register_node("default:chest_locked", {
+		local inv = meta:get_inventory()
+
+		if fields.rename or fields.key_enter_field == "text" then
+			meta:set_string("text", fields.text)
```
This should be limited to perhaps 1000 characters. I believe these nametags aren't supposed to be whole books.

SmallJoker commented on mods/default/nodes.lua:
```diff
@@ -1761,6 +1762,218 @@ minetest.register_node("default:chest_locked", {
+				new_stack:get_meta():set_string("description", fields.text)
+
+				minetest.log("action", sender:get_player_name().." renames "
+					..inv:get_stack("input", 1):get_name().." to "..new_stack:get_meta():get_string("description"))
```
Line too long

Ferk commented:
I'm not sure if it's wise to allow users to indiscriminately rename any item with any name.

Now that we have the metadata description feature there are a lot of possibilities that can be explored by mods using items with rich metadata, and being able to identify them using the description field.

If the description field can be easily overwritten by a workbench in MTG, then you will no longer be able to trust what the description says and it will lose its value.
How do you tell the difference between a sword that has been improved by a mod into a "+1 sword of strength" and an average sword from the same mod but that has been simply renamed in the workbench to have a "+1 sword of strength" description?

paramat commented:
We may need a group 'renamable' to limit what can be renamed.

paramat commented:
Yes not in nodes.lua.

C1ffisme commented:
> How do you tell the difference between a sword that has been improved by a mod into a "+1 sword of strength" and an average sword from the same mod but that has been simply renamed in the workbench to have a "+1 sword of strength" description?

As far as I know, you can't put colors into nametags. However, you can use minetest.colorize in the code, so that's an option.

Ferk commented:
> As far as I know, you can't put colors into nametags. However, you can use minetest.colorize in the code, so that's an option.

Does the description support colorization?

If the input field supports unicode, you might be able to copy-paste color sequences into it. [minetest.colorize](https://github.com/minetest/minetest/blob/c5967f75f0a9827d1b65b384edd6ba07c73ffd2f/builtin/game/misc.lua#L234-L244) is just a helper function to insert them.

DS-Minetest commented:
Perhaps just add the player given description to the original one instead of replacing it?

octacian commented:
* Moved to `workbench.lua`
* Limited description length to 500 characters, though I'm thinking 100 might be better
* Prevented "renaming" when the description is not changed (would waste a nametag if not prevented)
* Added check for `renameable` group, which if set to `0` in the item definition will prevent the item from being renamed (only the nametag uses this currently)
* Fixed line that was too long (mentioned by @SmallJoker)

Colours: Yes, you can put colours in nametags, though I haven't bothered with that here as I don't think it's important.

Just adding the player-given description to the original would really not be nice, plus it'd be harder to code. Being able to rename anything to anything (if it doesn't have `renameable = 0`) is way cooler IMO. Finally we can actually make up some good pranks in MT :P 

Anyways, I think this is ready to go aside from changing to a node, but it's unclear to me as to whether I need to. Personally, I prefer the complex nodebox, it adds some variety. At this point, I don't think I'll open a second PR using the crafting grid to perform considering the support this PR has received, that this is nicer in that it adds some variety (and challenge?), and considering that nobody (aside from sofar) has expressed that they'd prefer it the other way.

paramat commented:
So i assume nodes without a 'renameable' group will be renameable, seems ok i think.
I would prefer a node rather than this complex nodebox. Generally if something can reasonably be cubic then it should be a cube. Another reason is that smooth lighting on nodeboxes is quite intensive now.

Fixer-007 commented:
wait, may be better group nonrenameable, so only nodes that can't be renamed are marked in configs and everything else can be left as is?

paramat commented:
^ Possibly better, i have no opinion yet.

sofar commented:
I'm still very negative about this implementation.

I strongly feel we should do something similar as dyeing, and use the original craftgrid instead.

paramat commented:
I would be ok with a craft grid method: place tag and item in grid and rename, or maybe tag, renaming tool and item. Adding a workbench just for renaming seems odd i think i would prefer a dedicated tool.

C1ffisme commented:
> I'm still very negative about this implementation.

I strongly feel we should do something similar as dyeing, and use the original craftgrid instead.

I disagree. Having a portable workstation everywhere you go doesn't encourage the player to construct bases. Even for something as simple as renaming items. Besides, as mentioned earlier, it makes a good decoration block.

octacian commented:
I too, agree with @C1ffisme (I mean, ofc, otherwise I wouldn't have opened this PR :P)

Having these extra "little" things that could be expanded in the future adds more variety to MTG, more reason to settle down, a bit more of a challenge, and overall makes it feel more like a game. Not only that, but again, it's a cool decoration block. Just basic everything related to this on the crafting grid is just too easy. Plus, the only way to make that work would be if every renameable item had the renameable group set to `1`, which would be terrible IMO seeing as items should be renameable by default.

Anyways, I really hope that at least a few of the devs understand my point of view. MTG doesn't honestly deserve the "Game" ATM, this is something that would make it better.

sofar commented:
@C1ffisme minetest_game abandoned the workbench, so introducing it randomly for something that looks like a craft is not consistent. If we want that, we should introduce it for all crafting.

Besides, mods can change that behavior trivially.

C1ffisme commented:
@sofar 

> minetest_game abandoned the workbench, so introducing it randomly for something that looks like a craft is not consistent. If we want that, we should introduce it for all crafting.

Erm, Minetest Game never used a workbench in the first place, so you can't say we abandoned it.

I don't really understand what point you want to drive here. Do you want to change the name of the workbench? That can be done. Do you want it to look less like it's meant for crafting? That could also be done, although I like it's current look.

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,229 @@
+	local desc = stack:get_meta():get_string("description")
+
+	if desc == "" then
+		desc = minetest.registered_items[stack:get_name()].description
```
Potential crash with unknown items.

rubenwardy commented:
I like the idea of making usecase-specific craft nodes - such as anvils, smeltors, brewers, etc - for progression, and also allows moving away from boring 3x3 crafting. However, I'm unsure of whether a reduced version of this suits MTG. I feel like @paramat would say no :P

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,229 @@
+			return 0
+		elseif listname == "input" then
+			if minetest.registered_items[stack:get_name()].groups.renameable == 0 then
+        return 0
```
Your editor settings wrecked up somehow. Please look through the code and correct all of them (quite many).

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,229 @@
+		elseif listname == "output" then
+			return 0
+		elseif listname == "input" then
+			if minetest.registered_items[stack:get_name()].groups.renameable == 0 then
```
Again not safe from unknown items

rubenwardy commented:
> @rubenwardy This node does not add any API or changes the map generation, so I think it's fine to keep it in default.

It makes it considerably easier for server owners and other subgames to remove or modify it. It's a separate concept from the rest of default.

paramat commented:
> I feel like @paramat would say no :P

Hehe i'm undecided on the choice, i'm not really opposed to a node, as long as it is renamed for it't function as a name-stamper.

GreenXenith commented:
I like the nametag. Not the workbench. Unless the workbench is used for only renaming. We could copy MC and make an anvil that can rename :P

ForbiddenJ commented:
Is there a way to just append to the real item name like one of these? "New-Name (Real-Name)" "Real-Name (New-Name)"
Reason being, I have the technic mod, and the different uranium isotope items are visually the same. The only way to feasibly tell is via their names. If someone sells me a pile of 0.7% uranium ingots renamed to be "3.5%-Fissile Uranium Ingot" how can I quicly tell the difference before I actually try to use it?

octacian commented:
@ForbiddenJ I understand, however, if I required that the original name be kept somewhere in the new description, it would mean that for one thing many pranks would once again be impossible (e.g. you could prank someone with this uranium, renaming 3.5%-Fissile to 0.7%, which all though it sounds bad could be funny). When it comes to the renaming interface itself, I have coded this so that the original description automatically shows in the text field and can then be just added on to.

Now that I read your question, again, I understand your worry. Yes, it opens the way for pranks, but also for lying on the part of "shop owners" (though that is also true in the real world - someone can disguise something of low value as something of high value). I think the best solution for this would be to make a mod (since I don't think this would fit in MTG) that introduced an inspector which would compare the meta description with the original description.

I suggest the above because I personally often appreciate the ability to entirely change the name with no trace of the old.

Fixer-007 commented:
>If someone sells me a pile of 0.7% uranium ingots renamed to be "3.5%-Fissile Uranium Ingot" how can I quickly tell the difference before I try to use it?

By placing a block and reading its name in F5 mode.

octacian commented:
You can't place books in MTG.

Fixer-007 commented:
There are anti-prank workarounds: have at least one item/block you buy in your inventory, items you buy should stuck on it, if not - it is fraud.

ForbiddenJ commented:
@octacian As it is now, Minetest will tell one what the name is of an item honestly. It's like one can already tell what's what. Also, though Minecraft will let one rename an item, the game will italicize the name to tell that it has been renamed.

In the real world, people wouldn't know what an object is until they learn somehow, and thus they look at labels for an idea on what a product is. (What entity made it? What's it designed for? What features does it have? What is it?) Yes it can have false info, like there could be claims that say a light bulb will last longer than a competitor's, but one would know he's looking at a label attachment and not a game that doesn't tell him he's looking at one.

Now I do see where some pranks can be funny, like one can stick a diamond in a chest and then say "Hey! There's a diamond in that chest." and when someone steps near it, the floor opens up to a 5-node-deep hole. Or a [lapis luzuli](https://forum.minetest.net/viewtopic.php?f=11&t=11287&hilit=lapis) could be labeled "Diamond" and someone would say "hahaha! That's not a diamond!" That could either be hilarious, or bad. I hope for hilarious!
Just try to prevent the harmful ones. Okay?

I too would like to be able to attach information to items, so here's my suggestions:

- Add a color to the label to mean that it has been renamed (ex. orange for copper) along with a setting that will allow a server owner to specify which color labels will have, if any color. See the [enchanting](https://forum.minetest.net/viewtopic.php?f=11&t=14087) mod which is found in [X-Decor](https://forum.minetest.net/viewtopic.php?t=12534) to learn about coloring labels.

- Perhaps this PR as is could be used as a temporary April Fools joke?

- If all else fails, make it a separate mod for renaming. :)

C1ffisme commented:
If anything, I think that the chances of someone using item renaming to commit fraud are a bit slim, and you should be able to tell one item from another just by looking at the texture in vanilla Minetest. In modded Minetest, I think it would be good to add a group for items called `cannot_rename`, in the event that the item is, say, uranium fissile, and we don't want people confusing it with other fissile. (Because using 35 textures for uranium dust is a little bit silly.)

Fixer-007 commented:
>cannot_rename

thats already implemented

octacian commented:
Yes, the ability to prevent renaming has already been implemented as mentioned by @Fixer-007 

I am for the idea of colorizing the description, which I am already familiar with doing. Orange sounds good to me, though I wish we could do something more subtle like italics, but we can't at this point. I'm not sure about a settings to specify the colour though; devs here don't seem to really like those types of settings for whatever reason so I'd want some input on this from at least one of them.

ForbiddenJ commented:
I only suggested the color setting because you like subtle renames, but maybe the color can be left static. I prefer consistency anyway.
Aaaanyway, I like to make expensive tools (ex. mining lasers from [technic](https://forum.minetest.net/viewtopic.php?t=2538)). So maybe this feature would allow me to stick on a label that says "belongs to my_name".

paramat commented:
:-1:
The nodebox has 9 boxes (only 5 are necessary) i would prefer a normal cube because a workbench can very reasonably be a cube.
If we are going to add a 'workbench' with multiple uses then these multiple uses should be present from the start. Otherwise this should be named 'nameplater' or something similar.

Nodeboxes slow down rendering and need more intensive smooth lighting code, so it's important to resist the inevitable tendency for an increasing number of nodes being changed to nodeboxes. A voxel game represents non-cubic things as cubes to make things simple and lightweight, therefore there has to be a very good reason to use a nodebox, a workbench can obviously be a cube.

ForbiddenJ commented:
Oh man. I forgot about the workbench part of renaming. Is there a way to make this a craft-only thing and leave renaming a potential workbench feature should the workbench ever be added?

Maybe this could be done:

We can have a formspec like a sign, and acts like a book (saving, displaying, etc), but instead this would be on a name-tag, and then the name would reflect in tag's own name, then the tag can be crafted together with whatever else to be renamed. Special code for book copying can leave the original in place while copying with special code, and it looks like it can be redone for this PR.

To remove a tag, a renamed item would be crafted with a blank tag to remove this name using the same function.

Just an idea that might be a complete redo now that I think of it.

ForbiddenJ commented:
Or maybe you could just add the crafting grid for convenience, and to make it more meaningful.

DS-Minetest commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,229 @@
+			{-0.4375, -0.5, -0.4375, -0.25, 0.5, -0.25}, -- Leg2
+			{0.25, -0.5, 0.25, 0.4375, 0.5, 0.4375}, -- Leg3
+			{0.25, -0.5, -0.4375, 0.4375, 0.5, -0.25}, -- Leg4
+		}
```
This box is like this:
```
{-7/16, 3/16, -7/16, 7/16, 0.5, 7/16},
{-0.5, 1/4, -0.5, -7/16, 0.5, 0.5},
{7/16, 1/4, -0.5, 0.5, 0.5, 0.5},
{-0.5, 1/4, 7/16, 0.5, 0.5, 0.5},
{-0.5, 1/4, -0.5, 0.5, 0.5, -7/16},
{-7/16, -0.5, 1/4, -1/4, 0.5, 7/16},
{-7/16, -0.5, -7/16, -1/4, 0.5, -1/4},
{1/4, -0.5, 1/4, 7/16, 0.5, 7/16},
{1/4, -0.5, -7/16, 7/16, 0.5, -1/4}
```
This is too much. The same but better would be:
```
{-7/16, 3/16, -7/16, 7/16, 0.5, 7/16},
{-0.5, 1/4, -0.5, 0.5, 0.5, 0.5},
{-7/16, -0.5, 1/4, -1/4, 0.5, 7/16},
{-7/16, -0.5, -7/16, -1/4, 0.5, -1/4},
{1/4, -0.5, 1/4, 7/16, 0.5, 7/16},
{1/4, -0.5, -7/16, 7/16, 0.5, -1/4}
```
But even cheaper (and also nicer IMO) would be this:

```
{-0.5, 3/16, -0.5, 0.5, 0.5, 0.5},
{-7/16, -0.5, 1/4, -1/4, 0.5, 7/16},
{-7/16, -0.5, -7/16, -1/4, 0.5, -1/4},
{1/4, -0.5, 1/4, 7/16, 0.5, 7/16},
{1/4, -0.5, -7/16, 7/16, 0.5, -1/4}
```
(I've used HybridDog's nodebox_editor mod.)

SmallJoker replied:
This node isn't going to be used to build houses (hopefully), so some more nodeboxes to make it look fancy are ok.

DS-Minetest replied:
Actually IMHO the simpler version (3) looks even more fancy.
But 2 looks exactly the same as 1.

C1ffisme commented:
@paramat Minetest shouldn't be limited to simplistic, undetailed graphics just because we were too afraid that it would create lag. Plus, it's not like players are going to build houses out of these blocks.

ForbiddenJ commented:
Hey! Let's build a house entirely out of workbenches! LOL

Fixer-007 commented:
I doubt I will see a house out of workbenches, I don't care how it looks (at least now), only weird thing I seen is house out of chests... but that had protection reasons.

XtremeHacker commented:
I personally think the workbench could also be used for repairings items, by moving the repair type crafting recipes to be workbench only, and also if we want to add some sort of tool upgrade system, we could use the workbench to do suck things. :)

bigfoot547 commented:
Should you not add colorcodes so that you color the description? Also background color codes?
&0 should reset color
&1 red
&2 yellow
&21 revert red
&22 revert yellow
...
@octacian

benrob0329 commented:
+1 to this, and other machines in MTG. I think that more can be done with it in the future as well.

octacian commented:
@ForbiddenJ originally I wrote this with the idea of a nametag using the crafting grid just as you suggested (hence the branch name, `nametag`), however, I ran into a major issue. Essentially, to allow renaming ItemStacks in the crafting grid, each and every supported item would have to have a specific group (e.g. `renamable = 1`). IMO, the best way of going about this is to make it so that by default items can be renamed rather than require adding a group to nearly every item in MTG. This could be solved by looping through `minetest.registered_items`, however, it is an incredibly messy way of doing things. Now, some people don't agree with me that items should be renamable by default, but I personally believe that there would be more items for which one would want to support renaming then there would be for which one might want to disable renaming.

@SmallJoker I'll go through your requested changes as soon as I can.

@bigfoot547 IDK, colour codes seem a bit out of the scope of this PR considering that it would be inconsistent with things like chat, however, if it is brought to my attention that such a feature would be in the scope of this PR, I'd be happy to give it a go.

@DS-Minetest I'll tinker around with different node boxes and see what I can come up with, though I am a fan of adding more detail to MTG I must admit that my current node box is a bit over complex.

octacian commented:
@SmallJoker I believe I've fixed all issues you pointed out. The cause of formatting issues was the use of spaces in place of tabs (fixed). I've removed the possible crash due to unknown items by adding a simple check, `item:is_known()`, at the top of the two section which would be affected.

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,234 @@
+	local desc = stack:get_meta():get_string("description")
+
+	if desc == "" then
+		desc = minetest.registered_items[stack:get_name()].description
```
Could be returned directly

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,234 @@
+
+		if inv:is_empty("input") and inv:is_empty("nametag") and inv:is_empty("output") then
+			return true
+		else
```
`return inv:is_empty("input") and inv:is_empty("nametag") and inv:is_empty("output")`

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,234 @@
+				return 0
+			end
+		elseif to_list == "input" then
+			if minetest.registered_items[stack:get_name()].groups.renameable == 0 then
```
Apparently there's one more potential crash source.

octacian replied:
Where should the group be added in `game_api.txt`? I didn't find any other special groups like this, so not sure myself.

octacian replied:
Also, even if an item becomes undefined or is changed to `renameable = 0` while it is sitting in the workbench it will not crash the game (or did I misunderstand your question?).

Checks for unknown or renameable items are only made before the item is inserted or renamed (latter will soon be pushed). So, no, items can't be trapped in the slots.

SmallJoker replied:
@octacian What if the item becomes unknown when it's already in the slot?

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,234 @@
+					workbench_update_help(pos, nil, "Description too long (max 500 characters)!")
+					return
+				elseif fields.text == get_item_desc(inv:get_stack("input", 1)) then
+					workbench_update_help(pos, nil, "Description not changed!")
```
"Why was it not changed?"
IMO, you could simply return without throwing a message here since the item is named as wanted.

octacian replied:
I inserted this check as I thought that it'd be useless to "rename" an item if the new description was the same as the current description, therefore wasting a nametag. I can see that it could be confusing though, and maybe should be allowed.

SmallJoker replied:
That's a minor tweak. I don't mind if it's kept this way.

SmallJoker commented on mods/default/workbench.lua:
```diff
@@ -0,0 +1,234 @@
+
+				minetest.log("action", sender:get_player_name().." renames "
+					..inv:get_stack("input", 1):get_name().." to "
+						..new_stack:get_meta():get_string("description"))
```
`fields.text` is the same as the description meta. The variable could be used here to prevent some useless function calls.

paramat commented:
:-1: The nodebox has 9 boxes, that is excessive and out of place compared to our other simple nodebox nodes. You only need 1 for the top and 4 legs. I had not even noticed any additional boxes from looking at the screenshot, the difference will not be noticed unless you are under the table. Ideally i would prefer a normal cube because a workbench can very reasonably be a cube, just add transparent sections at the lower centre of each side face.

If we are going to add a 'workbench' with multiple uses then these multiple uses should be present from the start. Otherwise this should be named 'nameplater' or something similar.

C1ffisme commented:
@paramat

> The nodebox has 9 boxes, that is excessive and out of place compared to our other simple nodebox nodes. You only need 1 for the top and 4 legs. I had not even noticed any additional boxes from looking at the screenshot, the difference will not be noticed unless you are under the table.

This is almost a complete (Though, not literal) copy-paste of your previous comment. What is it with you and optimization? Are we not allowed to add features that may add a microsecond of lag? Because you must have turned a blind eye to TNT or view ranges of over 200 or so.

@paramat Just so I know what you are trying to go for, can you at least create a mockup version of what you want Minetest Game to look like? Because some of us are scratching our heads and beating them against the wall trying to figure that out.

paramat commented:
That's a vague and invalid argument, if you want to argue with me about my approach to this PR argue about specifics, i have clearly explained why 9 nobeboxes is excessive for MTG.
Obviously i am not opposed to features that add microseconds of lag if the extra processing is justified.
This is a case of design consistency.
Small optimisations add up to significant ones, MT is lightweight because it is the sum of thousands of unnoticeable optimisations, those who argue tha a single unnoticeable optimisation is insignificant don't see the bigger picture.

> Minetest shouldn't be limited to simplistic, undetailed graphics just because we were too afraid that it would create lag

MT isn't but MTG is graphically simple for good reason, the philosophy is simplicity and simple graphics for a lightweight subgame. If you want fancy graphics go play or create another subgame, we need more.

C1ffisme commented:
> Small optimisations add up to significant ones, MT is lightweight because it is the sum of thousands of unnoticeable optimisations, those who argue that a single unnoticeable optimisation is insignificant don't see the bigger picture.

But is it fun? We argue about single pixels on a texture, and we would be right in doing so! The tiniest, most insignificant graphical tweak or choice of background music can make a game fun or painful to play.

> MT isn't but MTG is graphically simple for good reason, the philosophy is simplicity and simple graphics for a lightweight subgame. If you want fancy graphics go play or create another subgame, we need more.

Minetest Game, being the default, non-third-party subgame, is like a sales pitch. If you want to bring in more people to enjoy the game, you need to make it look good, and deliver on it's promises. (None of this "Minetest has alien stuff" and then giving 1 block that has alien origins according to lore outside of the game itself.)

Think of Minetest Game as demonstrating the full power of the Minetest engine. If Minetest Game looks crappy, then what does that say about the engine?

XtremeHacker commented:
@C1ffisme Thing is the people using old hardware get upset then, you can never please everyone. :\
MTG is a mess though IMHO.

C1ffisme commented:
@XtremeHacker True.

DS-Minetest commented:
A mod could simply override the workbench node to make it to a complex nodebox. But nobody will have that much of an advantage because of a nodebox with 9 boxes (that are not even all needed).
IMHO a nodebox with 5 boxes or a full node would both be ok.

C1ffisme commented:
@DS-Minetest On the contrary, a mod could override the workbench node to have 5 boxes instead of 9.

octacian commented:
I've made every change requested by @SmallJoker except for adding information about the `renameable` group to `game_api.txt` as I don't know how or where it should be added. I also chose to use the nodebox with the smallest number of boxes, provided by  @DS-Minetest, as I thought it looked the nicest while still saving on performance.

I really want to see this in 0.4.16 (for one thing, it'll give me more to talk about in my video :rotfl:), I'd be happy to rename the node to something like a "Name Plater" if it is wanted.

Please re-review.

paramat commented:
Thanks, better, but i prefer a node instead of a nodebox. More importantly, this should perhaps be a separate mod in MTG.
There is still controversy and this seems unsettled: is this going to be a multipurpose workbench or a one-use nameplater? If it's only a nameplater should it just be a simple tool instead? So this is nowhere near ready for release.
:-1:

C1ffisme commented:
@paramat 
> Thanks, better, but i prefer a node instead of a nodebox

Again I say, nobody is going to build a house out of these blocks.

We've had a mod API for the last 5 years, are we not allowed to use it to it's full potential just because we need to shave off nanoseconds of possible lag?

Congratulations! We've made a game that runs at 120 FPS that looks too ugly to stand for more than 5 seconds and isn't fun to play at all!

bigfoot547 commented:
We should get rid of the stairs and slabs mesh too!

XtremeHacker commented:
And we may as well make the player a node...

paramat commented:
That's just my preference, my -1 is more for other reasons.

> We've had a mod API for the last 5 years, are we not allowed to use it to it's full potential just because we need to shave off nanoseconds of possible lag?
Congratulations! We've made a game that runs at 120 FPS that looks too ugly to stand for more than 5 seconds and isn't fun to play at all!

It's not about saving a tiny amount of rendering time.
The second part of your comment therefore applies to how MTGame is now, so why do you play it at all or comment here?
Typically silly comments from you.

C1ffisme commented:
> It's not about saving a tiny amount of rendering time.

Then why do it? What's the point?

> The second part of your comment therefore applies to how MTGame is now without a nodebox workbench, so why do you play it at all or comment here?

Mostly that was my response to "the game if we keep making decisions like this". Lemme edit for clarity.

> You've reduced the amount of comments you make, which is welcome, but have not increased the quality.

I have seen comments on Github made by people who have submitted and merged pull requests to this repository which look and read far worse than mine.

paramat commented:
> Then why do it? What's the point?

Explained above, try reading.

> We've had a mod API for the last 5 years, are we not allowed to use it to it's full potential just because we need to shave off nanoseconds of possible lag?
(The game if we keep doing this:)
Congratulations! We've made a game that runs at 120 FPS that looks too ugly to stand for more than 5 seconds and isn't fun to play at all!

'keep doing this' refers to not adding new nodebox content, it does not refer to removing content, so your comment refers to how MTGame is at the moment

> I have seen comments on Github made by people who have submitted and merged pull requests to this repository which look and read far worse than mine.

But at least they have contributed, so some bad comments are tolerated, You, along with one other, had to be warned about a high amount of worthless comments.

C1ffisme commented:
> But at least they have contributed

[And](https://github.com/minetest/minetest_game/pull/1495) [I've](https://github.com/minetest/minetest_game/pull/1032) [tried](https://github.com/minetest/minetest_game/pull/1007) [but](https://github.com/minetest/minetest_game/pull/447) [I'm just somewhat bad at contributing code.](https://github.com/minetest/minetest_game/pull/444)

[(I have, however, made one successful suggestion that has benefited the game.)](https://github.com/minetest/minetest_game/issues/738)

All I've gotten from trying to contribute code was de-motivation and inhospitality. If all I can contribute is advice, which I've gathered from playing other, similar games for just a bit longer than is healthy, then I'll try and do that.

> 'keep doing this' refers to not adding new nodebox content, it does not refer to removing content, so your comment refers to how MTGame is at the moment

I'm not sure how you're reading me, and that statement was mostly humorous, so I'm just going to give up on this.

Just so you know, I've played Minetest primarily from 2012 to 2014, and I've played mostly Minecraft up to today, with a bit of Terraria and Starbound here and there. (Before 2012, I also played a few other open-source games, as well as some of the Age of Empires games and Empire Earth.) I have quite a bit of free time, and I've accumulated a lot of knowledge and opinions on what can make a game feel enjoyable, and addictive, and what can make a game feel convoluted and ugly. I wouldn't consider myself a professional, but I can say I've had experience with good games.

I've seen your mods, and their artwork, and I know you have a certain taste in game art. (The "OCD, flat style") But from what I've seen in the gaming world, it's not an entirely objective one.

paramat commented:
Sorry, i see you have contributed.

C1ffisme commented:
@paramat It's okay. You're forgiven. :wink:

paramat commented:
> and that statement was mostly humorous,

I'm sorry, i have been over-argumentative and pedantic in this thread.

DS-Minetest commented:
The simplest way to implement this could be with a chatcommand (that needs a special priv or just give).

Fixer-007 commented:
I guess moving from tool repair from grid to workbench is a good idea.

paramat commented:
> The simplest way to implement this could be with a chatcommand

That seems more suitable for this single feature of renaming, or maybe a tool, or using the crafting grid somehow. Renaming is a minor feature and a workbench node is overkill for just that.

If we have a workbench it needs to come with multiple uses (and is therefore justified) from the start. I know that more uses are intended, but that may never happen, and if more uses are intended then they can be coded from the start.

DS-Minetest commented:
If a workbench would really be added, perhaps an api would be nice for mods to add more tabs (like in player inventory I think).

paramat commented:
> This should be a new mod,

I agree, we have had many requests to add new features in separate mods if at all possible.

This PR is very controversial, with 2 disapprovals. Will close in a month if no progress.

octacian commented:
@paramat I plan to work on this feature in the beginning of October (will be back to MT then). I'm happy to modify it however needed in order to get it merged, but I strongly believe we are overusing the crafting grid and would be better off to use more creative methods of achieving these things. Also, renaming in the crafting grid would be a bad idea AFAIK. Pretty sure that any item for which we wanted to allow renaming would have to have a group set which would require that we add the group to almost everything in MTG, as there are very few items that *shouldn't* be renamable. Unless it's possible to only apply in the crafting grid if the item *doesn't* have a particular group rather than only if it *does*, the workbench is a better option for yet another reason. I may be wrong though, will have to check up on this. Anyways, assuming there is still interest in the workbench method, I'll detail what I was going to do in order to make the workbench more useful.

Aside from any basic bugfixes and improvements, I'd like to move tool repair from the crafting grid to the workbench. This would allow for two damaged tools to be placed in each slot, the resulting item being the combination of the remaining wear as is currently done in the crafting grid. Otherwise, each tool could be repaired by inserting the damaged tool and the material used to make the tool. For example, a diamond pick could be repaired 50% with one diamond, a diamond sword could be repaired ~60% with one diamond. In case you wonder why the sword, for example, wouldn't be repaired 50% with one diamond, it's because that would mean that a full repair would cost two diamonds, the same amount required to make a new sword, taking away any advantages that could be gained by using the workbench.

With the above combined with already implemented renaming features, I think the workbench would become much more worthwhile to add to MTG.

paramat commented:
Yeah sounds like that has potential. It really needs to be a separate mod and with all the features present to justify a workbench.
I might be wrong that naming items can be done in the crafting grid, i need to reread the discussion.

paramat commented:
Neglected long term, can reopen on progress.

Fixer-007 commented:
Now, since it is neglected and some more time passed I have alternative idea, instead of this C O N G L O M E R A T E, that spins out of control asking very paramat-unfriendly questions like maybe we should have workbench for crafting or repair and items should not stay in player's crafting grid (seriously, anyone remembers this silliness!?) and if it can't, then items are dropped, then autopickup/notminetoinventory debate starts, it is a big mess, here is simple alternative:

Introducing "Seal", craftable disposable item, workflow: craft seal, right click - formspec asks for seal text - done - apply on any block/dropped item by aiming with it and left-clicking (or combining in inventory or both), seal gets consumed renaming pointed block in the process.

paramat commented:
The author seems to have moved on, but we may still be in need of a renaming feature, but simple, like a tool used in the crafting grid.
